haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs

```diff
@@ -220,8 +220,22 @@ declsFromRenamed ctx (hsGroup, _, _, _) =
                                 $ hsGroup
         explicitVarBindDefs = map mkDeclName explicitNames
         varBindDefs = explicitVarBindDefs ++ implicitVarBindDefs
-    in (concat [defs, instDefs, varBindDefs], instChanges)
+        valBindDefs = case hs_valds hsGroup of
+            ValBinds _ _ _ ->
+                error "should not hit ValBinds when accessing renamed AST"
+            XValBindsLR (NValBinds binds _sigs) -> collectPatSymDefs binds
+    in (concat [defs, instDefs, varBindDefs, valBindDefs], instChanges)
   where
+    -- TODO(jinwoo): Collect for signatures too.
+    collectPatSymDefs :: [(GHC.RecFlag, LHsBinds GhcRn)] -> [DeclAndAlt]
+    collectPatSymDefs binds = concatMap collectPatBinds binds
+    collectPatBinds :: (GHC.RecFlag, LHsBinds GhcRn) -> [DeclAndAlt]
+    collectPatBinds (_, bindsBag) =
+      mapMaybe collect (GHC.bagToList bindsBag)
+      where
+        collect (L _ (PatSynBind _ PSB {psb_id = psbId})) =
+          Just $ declWithWrappedIdLoc (outputableStringyType ctx psbId) psbId
+        collect _ = Nothing
     keepFirst [] = Nothing
     keepFirst xs@((name, _):_) = Just $! (name, minimum $ map snd xs)
```

Comment:
I know this is pre-existing, but just in case you feel like changing this:
(,) has exactly one constructor, so the extra strictness of $! is a no-op.
`Just (name, minimum $ map snd xs)` is just as good.

Reply:
I'm fine either way, but wouldn't `$!` save a thunk? Ok, maybe that's premature optimization, but I was deeply touched by @tibbe's (not very recent) Haskell performance slides. What are the general guidelines nowadays?